near-sdk/src/json_types/public_key.rs

```diff
@@ -132,24 +170,22 @@ impl std::str::FromStr for Base58PublicKey {
     type Err = ParsePublicKeyError;
 
     fn from_str(value: &str) -> Result<Self, Self::Err> {
-        let (key_type, key_data) = Base58PublicKey::split_key_type_data(&value)?;
-        let expected_length = match key_type {
-            CurveType::ED25519 => 32,
-            CurveType::SECP256K1 => 64,
-        };
+        let (curve, key_data) = Base58PublicKey::split_key_type_data(&value)?;
         let data = bs58::decode(key_data).into_vec()?;
-        if data.len() != expected_length {
-            return Err(ParsePublicKeyError {
-                kind: ParsePublicKeyErrorKind::InvalidLength(data.len()),
-            });
-        }
-        let mut res = Vec::with_capacity(1 + expected_length);
-        match key_type {
-            CurveType::ED25519 => res.push(0),
-            CurveType::SECP256K1 => res.push(1),
-        };
-        res.extend(data);
-        Ok(Self(res))
+        Self::from_parts(curve, data)
+    }
+}
+
+impl BorshDeserialize for Base58PublicKey {
+    fn deserialize(buf: &mut &[u8]) -> io::Result<Self> {
+        Ok(Self { curve: CurveType::deserialize(buf)?, data: BorshDeserialize::deserialize(buf)? })
+    }
+}
+
+impl BorshSerialize for Base58PublicKey {
+    fn serialize<W: io::Write>(&self, writer: &mut W) -> io::Result<()> {
+        self.curve.serialize(writer)?;
+        BorshSerialize::serialize(&self.data, writer)
     }
 }
```

Comment:
unless I'm mistaken about borsh internals, this is just doing what the derive is doing and will be in the wrong format still. Can double-check in a bit, but I don't believe this will be correct.

Reply:
What would you recommend to change it to? I'm not sure what the right format would be

Reply:
the format previously would just be serializing as a single `Vec<u8>` of bytes. Try creating a serialized vector from what exists now with borsh and match with what's changed now.

I'm 90% sure borsh derive just serializes as the inner newtype, which previously was `Vec<u8>`, but best to double check.

Reply:
Oh I see what you mean. I'm not too familiar with how borsh/serde does (de)serialization. I'm guessing there's not an equivalent to versioning like how protobuf does it, since we gotta manually serialize it the same way as a previous version

Reply:
Yeah, we wouldn't want it versioned because then more data than necessary would need to be stored